lxd-agent/network.go

```diff
@@ -69,3 +78,101 @@ func serverTLSConfig() (*tls.Config, error) {
 
 	return tlsConfig, nil
 }
+
+// reconfigureNetworkInterfaces checks for the existence of "nics.json" in the config share.
+// Interfaces are identified by their MAC address and configured accordingly.
+func reconfigureNetworkInterfaces() {
+	// Abort if configuration file does not exist (nothing to do), otherwise log and return.
+	nicDir, err := ioutil.ReadDir("nics")
+	if os.IsNotExist(err) {
+		return
+	} else if err != nil {
+		logger.Errorf("Could not read network interface configuration directory: %s", err.Error())
+		return
+	}
+
+	nicData := make(map[string]deviceConfig.NICConfig)
+	for _, f := range nicDir {
+		nicBytes, err := ioutil.ReadFile(filepath.Join("nics", f.Name()))
+		if err != nil {
+			logger.Errorf("Could not read network interface configuration file: %s", err.Error())
+		}
+
+		var conf deviceConfig.NICConfig
+		err = json.Unmarshal(nicBytes, &conf)
+		if err != nil {
+			logger.Errorf("Could not parse network interface configuration file: %s", err.Error())
+			return
+		}
+
+		nicData[f.Name()] = conf
+	}
+
+	netDir := "/sys/class/net"
+	files, err := ioutil.ReadDir(netDir)
+	if err != nil {
+		return
+	}
+
+	reverter := revert.New()
+	for _, f := range files {
+		addrBytes, err := ioutil.ReadFile(filepath.Join(netDir, f.Name(), "address"))
+		if err != nil {
+			return
+		}
+		macAddress := string(bytes.TrimRight(addrBytes, "\n"))
+
+		mtuBytes, err := ioutil.ReadFile(filepath.Join(netDir, f.Name(), "mtu"))
+		if err != nil {
+			return
+		}
+		mtuStr := string(bytes.TrimRight(mtuBytes, "\n"))
+
+		nic, ok := nicData[macAddress]
```

Comment:
This doesn't look right, earlier on line 108 the nicData was keyed on file name not mac address.

Reply:
Yes on line 108 the file name is the MAC address (see https://github.com/lxc/lxd/pull/9893/files#diff-87040b4a9b898c058e0f889f869c7a0448135f0a7f02db851824f48290c3a2f5R3078). So we have `/run/lxd_agent/nics/<mac address>` files containing the json for the device name and MTU.

Then we find the interface of interest by looking for the `/sys/class/net/<interface>/address` which matches that file name.

Reply:
In any case this is clearly confusing. I can name the files by the device name (i.e. `lxdbr0.json`) and also include MAC address in the `NicConfig`?

Reply:
This would also align nicely with how you named the dnsmasq config files in https://github.com/lxc/lxd/pull/9790

Reply:
Yep sounds good :+1:

Reply:
Oh whilst you're at it, can you check the validation for specifying the device name (not the interface name), as I think this PR and https://github.com/lxc/lxd/pull/9790 are the first time we've started using device name as a file on the filesystem and want to check things like `.`, `/` and `..` aren't allowed to avoid the chance of writing these files in the wrong dir.

Reply:
We will probably want a test for it in case it gets regressed in the future too.

Reply:
As far as I can tell there is no validation on device names. Currently both `.` and `/` are allowed characters. They both will break the dnsmasq implementation:
```
$ lxc config device add c1 ../n1 nic nictype=bridged parent=lxdbr2
Error: Failed to add device "../n1": open /var/lib/lxd/networks/lxdbr2/dnsmasq.hosts/c1.../n1: no such file or directory
```

Reply:
I think we will need to disallow both `/` and `.` from device names.

Reply:
@stgraber shall we limit device names to valid hostname chars plus underscore?

Reply:
Yeah, that sounds good.

Reply:
Thanks. @markylaing can you send up a PR for the dnsmasq fix so it can be cherry picked into 4.22 please .

Reply:
@markylaing so we want something like https://github.com/lxc/lxd/blob/53faba666fc0881ab8679b4c6608fdb474da0b0d/shared/util.go#L724-L751 but that it also allows underscores. It should be called from `Validate` (https://github.com/lxc/lxd/blob/a777a856b1a114a0d3bb30b0cc385289fb2b8699/lxd/device/device_load.go#L128) and `New` (https://github.com/lxc/lxd/blob/a777a856b1a114a0d3bb30b0cc385289fb2b8699/lxd/device/device_load.go#L111) and be saved in `device_load.go`. Thanks

Reply:
I've added some commits for this now.